pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go

```diff
@@ -99,6 +113,111 @@ type HTTPConnectionSettings struct {
+	// host without any involvement of a global rate limiting service.
+	// This is applied as a token bucket rate limiter.
+	// +optional
+	Local *LocalRateLimitSpec `json:"local,omitempty"`
+}
+
+// LocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host.
+type LocalRateLimitSpec struct {
+	// TCP defines the local rate limiting specification at the network
+	// level. This is a token bucket rate limiter where each connection
+	// consumes a single token. If the token is available, the connection
+	// will be allowed. If no tokens are available, the connection will be
+	// immediately closed.
+	// +optional
+	TCP *TCPLocalRateLimitSpec `json:"tcp,omitempty"`
+
+	// HTTP defines the local rate limiting specification for HTTP traffic.
+	// This is a token bucket rate limiter where each request consumes
+	// a single token. If the token is available, the request will be
+	// allowed. If no tokens are available, the request will receive the
+	// configured rate limit status.
+	HTTP *HTTPLocalRateLimitSpec `json:"http,omitempty"`
+}
+
+// TCPLocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host at the TCP level.
+type TCPLocalRateLimitSpec struct {
+	// Connections defines the number of connections allowed
+	// per unit of time before rate limiting occurs.
+	Connections uint32 `json:"connections"`
+
+	// Unit defines the period of time within which connections
+	// over the limit will be rate limited.
+	// Valid values are "second", "minute" and "hour".
+	Unit string `json:"unit"`
+
+	// Burst defines the number of connections above the baseline
+	// rate that are allowed in a short period of time.
+	// +optional
+	Burst uint32 `json:"burst,omitempty"`
+}
+
+// HTTPLocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host at the HTTP level.
+type HTTPLocalRateLimitSpec struct {
+	// Requests defines the number of requests allowed
+	// per unit of time before rate limiting occurs.
+	Requests uint32 `json:"requests"`
+
+	// Unit defines the period of time within which requests
+	// over the limit will be rate limited.
+	// Valid values are "second", "minute" and "hour".
+	Unit string `json:"unit"`
+
+	// Burst defines the number of requests above the baseline
+	// rate that are allowed in a short period of time.
+	// +optional
+	Burst uint32 `json:"burst,omitempty"`
+
+	// ResponseStatusCode defines the HTTP status code to use for responses
+	// to rate limited requests. Code must be in the 400-599 (inclusive)
+	// error range. If not specified, a default of 429 (Too Many Requests) is used.
+	// +optional
+	ResponseStatusCode uint32 `json:"responseStatusCode,omitempty"`
+
+	// ResponseHeadersToAdd defines the list of HTTP headers that should be
+	// added to each response for requests that have been rate limited.
+	// +optional
+	ResponseHeadersToAdd []HTTPHeaderValue `json:"responseHeadersToAdd,omitempty"`
```

Comment:
Is there any value in this not just being a `map[string]string`?

Reply:
Ah, good call. Do we need to worry about this one too?

> Do not use unsigned integers, due to inconsistent support across languages and libraries. Just validate that the integer is non-negative if that's the case.

Reply:
I think it's fine here, as we aren't really in the business of supporting multiple clients for different languages similar to the k8s project. We are only publishing the Golang APIs for our own control plane to consume. I have seen uint being used similarly across other projects.